### PR TITLE
[REVIEW] Replace CUDA_RT_CALL with CUDA_TRY

### LINKED ISSUE #587: CUDA_RT_CALL macro should not call exit()
The CUDA_RT_CALL macro in src/hash-join/concurrent_unordered_multimap.h calls exit(1) when an error is encountered. This could lead to memory leaks since this gives no opportunity to clean up memory. Instead of exiting, the macro should propagate the error (return an error, or throw, as appropriate).

### PR DESCRIPTION
**Pull request by seunghwak**
This PR addresses https://github.com/rapidsai/cudf/issues/587 and replaces CUDA_RT_CALL (which calls exit()) with CUDA_TRY (which throws).

Fixes #587

### CODE REVIEW

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 1]
```diff
@@ -17,22 +17,22 @@
 #ifndef CONCURRENT_UNORDERED_MULTIMAP_CUH
 #define CONCURRENT_UNORDERED_MULTIMAP_CUH
 
+#include "managed.cuh"
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]:
```suggestion
#include <hash/managed.cuh>
```

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 2]
```diff
@@ -17,22 +17,22 @@
...
+#include "hash_allocator.cuh"
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]:
```suggestion
#include <hash/hash_allocator.cuh>
```

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 3]
```diff
@@ -17,22 +17,22 @@
...
+#include "helper_functions.cuh"
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]:
```suggestion
#include <hash/helper_functions.cuh>
```

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 4]
```diff
@@ -17,22 +17,22 @@
...
+#include <utilities/legacy/device_atomics.cuh>
+
 #include <cudf/cudf.h>
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Don't think we need this anymore

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 5]
```diff
@@ -18,11 +18,14 @@
 #define CONCURRENT_UNORDERED_MAP_CUH
 
 #include <utilities/legacy/device_atomics.cuh>
-#include <cudf/detail/utilities/hash_functions.cuh>
 #include "helper_functions.cuh"
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_map.cuh]: #include <hash/helper_functions.cuh>

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 6]
```diff
@@ -18,11 +18,14 @@
...
 #include "hash_allocator.cuh"
```
Comment (rgsl888prabhu) [inline cpp/src/hash/concurrent_unordered_map.cuh]: #include <hash/hash_allocator.cuh>

File: cpp/src/hash/concurrent_unordered_map.cuh [hunk 7]
```diff
@@ -422,16 +425,16 @@ class concurrent_unordered_map {
...
             m_hashtbl_values, m_capacity * sizeof(value_type), dev_id, stream));
       }
     }
 
     init_hashtbl<<<((m_capacity - 1) / block_size) + 1, block_size, 0, stream>>>(
         m_hashtbl_values, m_capacity, m_unused_key, m_unused_element);
-    CUDA_RT_CALL(cudaGetLastError());
-    CUDA_RT_CALL(cudaStreamSynchronize(stream));
+    CUDA_TRY(cudaGetLastError());
+    CUDA_TRY(cudaStreamSynchronize(stream));
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_map.cuh]: Is the synchronize required here? As long as the hash table is later used in the same stream as it is initialized on it shouldn't be, right? @nsakharnykh @jrhemstad thoughts?

File: cpp/src/hash/concurrent_unordered_multimap.cuh [hunk 8]
```diff
@@ -502,17 +501,17 @@ class concurrent_unordered_multimap {
...
       }
     }
 
     if (init) {
       init_hashtbl<<<((m_hashtbl_size - 1) / block_size) + 1, block_size, 0, stream>>>(
           m_hashtbl_values, m_hashtbl_size, unused_key, unused_element);
-      CUDA_RT_CALL(cudaGetLastError());
-      CUDA_RT_CALL(cudaStreamSynchronize(stream));
+      CUDA_TRY(cudaGetLastError());
+      CUDA_TRY(cudaStreamSynchronize(stream));
```
Comment (harrism) [inline cpp/src/hash/concurrent_unordered_multimap.cuh]: Same question as above.

File: cpp/src/hash/concurrent_unordered_map.cuh (hunk 7 above)
Comment (nsakharnykh) [reply cpp/src/hash/concurrent_unordered_map.cuh]: The synchronization is not required here and should be removed. In fact, in some other uses of the hash maps, I remember I had to remove it and keep a separate modified copy of cuDF. It’s a better design to let users synchronize outside the class if necessary.

Comment (seunghwak) [reply cpp/src/hash/concurrent_unordered_map.cuh]: OK, I will remove the cudaStreamSynchronize calls.

Comment (seunghwak) [reply cpp/src/hash/concurrent_unordered_map.cuh]: It seems like just removing cudaStreamSynchronize is not a proper resolution. If I am removing synchronization here, I should add something elsewhere or need to check whether all the use cases are properly using stream.

For example, we are taking stream in the function below,

https://github.com/rapidsai/cudf/blob/f8c1c0d5e5336d2890ca1a6f2ec9bedad5d4694d/cpp/src/groupby/hash/legacy/groupby.cu#L53

but, we are not passing stream to the create function here (which calls cudaStreamSynchronize at the end)

https://github.com/rapidsai/cudf/blob/f8c1c0d5e5336d2890ca1a6f2ec9bedad5d4694d/cpp/src/groupby/hash/legacy/groupby.cu#L92

but, using stream when using map returned by the create function.

https://github.com/rapidsai/cudf/blob/f8c1c0d5e5336d2890ca1a6f2ec9bedad5d4694d/cpp/src/groupby/hash/legacy/groupby.cu#L101

I will dig deeper to check what else I need to do to safely remove synchronization, here.

And we're still using legacy code, so I assume we cannot legacy code yet, but please let me know if I am doing something unnecessary.

Comment (harrism) [reply cpp/src/hash/concurrent_unordered_map.cuh]: The port of groupby will use the hash maps.  It is not ported yet. I think the groupby port should fix this issue @jrhemstad

Comment (seunghwak) [reply cpp/src/hash/concurrent_unordered_map.cuh]: OK, I will revert the last two commits expecting this issue to be addressed elsewhere.

Comment (harrism) [reply cpp/src/hash/concurrent_unordered_map.cuh]: Oh, I misunderstood -- I thought you were reporting a problem. I didn't know you had already fixed it. Why revert it?

Comment (seunghwak) [reply cpp/src/hash/concurrent_unordered_map.cuh]: OK, re-applied the changes.

**Pull request merged**